### DISCUSSION TITLE
Migrating an existing cluster's nodes to Talos

### DISCUSSION THREAD

**Discussion by ethernet-zero** [siderolabs/talos#5644] [Q&A] [answered]
Hi!

Would it be possible to migrate an entire cluster that's already been set up with Kubespray on Flatcar nodes? I mean, can you join a Talos node to an existing cluster that's not been set up with Talos, or even replace the controlplane nodes one by one?

My guess is that, as long as you provide the right configuration corresponding to the existing cluster, it should probably work, right?

**Comment by smira** [comment]
It's currently possible to join non-Talos worker nodes to the Talos control plane, but vice versa won't work. Talos API on worker nodes requires a connection to `trustd` component which only exists on Talos control plane nodes.

**Comment by ethernet-zero** [comment]
And what about control plane nodes? Would it be possible to create a Talos control plane node but make it join an existing control plane instead of boostrapping the cluster from scratch? If it could be done, we could just replace nodes in the control plane one by one and then do the same with all the workers, since there would be a full Talos control plane by then.

**Reply by smira** [reply]
This should probably work if all secrets match, but I never tried that, so :man_shrugging:

**Reply by ethernet-zero** [reply]
I feel adventurous. :grin: 

I'll post my findings here when I have the opportunity to test it.

**Reply by PhilippKuntschik** [reply]
@ethernet-zero did you test adding a talos control plane node to an existing cluster?

**Reply by ethernet-zero** [reply]
Hi, I'm finally going to test it this week, using the guide posted by @smira. I'll add my comments on the process when I'm done, I'll test with a small cluster: first migrating it to Talos, then waiting for upgrades and upgrading it, and probably also adding a couple of workers to it.

**Comment by smira** [comment, marked as answer]
We have a way now with 1.2.0 release, a guide https://www.talos.dev/v1.2/advanced/migrating-from-kubeadm/

**Comment by ethernet-zero** [comment]
Hi,

The guide mentions setting the CNI plugin name to `none`. My guess is so the installation doesn't affect the CNI plugin currently used by the cluster, is that right? So, it's a matter of letting it join the cluster and it will schedule the Calico nodeplugin normally then?

Same thing with the external cloud controller. If I don't define any in the `controlplane.yaml` file, it should just schedule the node controllers and pick up the cloud config from the ConfigMap, correct?

**Reply by smira** [reply]
CNI `none` means that Talos won't try to deploy any CNI manifests.

As CNI is already deployed to the cluster, that is fine.

**Comment by ethernet-zero** [comment]
I got the first Talos control plane node to join the etcd cluster and start initializing pods, but now `kube-apiserver` doesn't become reachable by the controller-runtime because it has trouble trying to contact the `metrics-server.svc` service, I guess because Calico is not yet running because `kube-apiserver` is not running yet. Here are some log messages:

```
talos-anonymized-node-name: 2022-09-21T10:15:26.159069674Z stderr F E0921 10:15:26.159004       1 controller.go:116] loading OpenAPI spec for "v1beta1.metrics.k8s.io" failed with: failed to retrieve openAPI spec, http error: ResponseCode: 503, Body: error trying to reach service: dial tcp <METRICS-SERVER_SVC_IP>:443: i/o timeout
talos-anonymized-node-name: 2022-09-21T10:15:26.159090458Z stderr F , Header: map[Content-Type:[text/plain; charset=utf-8] X-Content-Type-Options:[nosniff]]
talos-anonymized-node-name: 2022-09-21T10:15:26.159142181Z stderr F I0921 10:15:26.159023       1 controller.go:129] OpenAPI AggregationController: action for item v1beta1.metrics.k8s.io: Rate Limited Requeue.
talos-anonymized-node-name: 2022-09-21T10:15:30.224287246Z stderr F E0921 10:15:30.224203       1 available_controller.go:524] v1beta1.metrics.k8s.io failed with: Operation cannot be fulfilled on apiservices.apiregistration.k8s.io "v1beta1.metrics.k8s.io": the object has been modified; please apply your changes to the latest version and try again
talos-anonymized-node-name: 2022-09-21T10:15:35.236581735Z stderr F E0921 10:15:35.236435       1 available_controller.go:524] v1beta1.metrics.k8s.io failed with: failing or missing response from https://<METRICS-SERVER_SVC_IP>:443/apis/metrics.k8s.io/v1beta1: Get "https://<METRICS-SERVER_SVC_IP>:443/apis/metrics.k8s.io/v1beta1": net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
talos-anonymized-node-name: 2022-09-21T10:15:40.259410705Z stderr F E0921 10:15:40.259265       1 available_controller.go:524] v1beta1.metrics.k8s.io failed with: Operation cannot be fulfilled on apiservices.apiregistration.k8s.io "v1beta1.metrics.k8s.io": the object has been modified; please apply your changes to the latest version and try again
```

Does it make any sense that this could have hit a chicken/egg situation?

**Reply by smira** [reply]
I'm not sure what you mean by that. If you have running existing cluster, it should have apiserver running already. Talos nodes will talk to the existing control plane endpoint.

**Reply by ethernet-zero** [reply]
Well, maybe I'm confusing things. What I meant is I found this in the dmesg logs:

```
[ 1983.102429] [talos] controller failed {"component": "controller-runtime", "controller": "k8s.KubeletStaticPodController", "error": "error refreshing pod status: error fetching pod status: an error on the server (\"Authorization error (user=apiserver-kubelet-client, verb=get, resource=nodes, subresource=proxy)\") has prevented the request from succeeding"}
```

And according to [the documentation](https://www.talos.dev/v1.2/advanced/troubleshooting-control-plane/#talos-prints-error-an-error-on-the-server--has-prevented-the-request-from-succeeding), this is what it means:

> k8s.KubeletStaticPodController: controller failed: error refreshing pod status: error fetching pod status: an error on the server ("Authorization error (user=apiserver-kubelet-client, verb=get, resource=nodes, subresource=proxy)") has prevented the request from succeeding: kubelet hasn’t been able to contact kube-apiserver yet to push pod status, controller is going to retry.

And I'm not sure if it's related, but I guess that could be at least part of the cause why the node is not listed in `kubectl get nodes` although its endpoints show up in `kubectl get endpoints` and why the `calico-node` DaemonSet is not even attempting to schedule a pod in the new node, which maybe is why the new node can't contact any pods or services.

I'm pretty sure it's just some silly thing I missed that, when fixed, will make everything fall into place.

**Reply by smira** [reply]
the error itself usually means that kubelet hasn't been able to join the controlplane yet.

this doesn't show the root cause unfortunately, so you might need to check `kubelet` logs

**Reply by ethernet-zero** [reply]
Ah, there you go:

```
talos-anonymized-node-name: {"ts":1663845935234.3503,"caller":"lease/controller.go:144","msg":"failed to ensure lease exists, will retry in 7s, error: leases.coordination.k8s.io \"talos-anonymized-node-name\" is forbidden: User \"system:anonymous\" cannot get resource \"leases\" in API group \"coordination.k8s.io\" in the namespace \"kube-node-lease\"\n"}
talos-anonymized-node-name: {"ts":1663845935265.428,"caller":"kubelet/kubelet.go:2466","msg":"Error getting node","err":"node \"talos-anonymized-node-name\" not found"}
talos-anonymized-node-name: {"ts":1663845935366.2693,"caller":"kubelet/kubelet.go:2466","msg":"Error getting node","err":"node \"talos-anonymized-node-name\" not found"}
talos-anonymized-node-name: {"ts":1663845935442.4597,"caller":"kubelet/kubelet_node_status.go:70","msg":"Attempting to register node","v":0,"node":{"name":"talos-anonymized-node-name"}}
talos-anonymized-node-name: {"ts":1663845935461.8638,"caller":"kubelet/kubelet_node_status.go:92","msg":"Unable to register node with API server","node":{"name":"talos-anonymized-node-name"},"err":"nodes is forbidden: User \"system:anonymous\" cannot create resource \"nodes\" in API group \"\" at the cluster scope"}
talos-anonymized-node-name: {"ts":1663845935467.1292,"caller":"kubelet/kubelet.go:2466","msg":"Error getting node","err":"node \"talos-anonymized-node-name\" not found"}
talos-anonymized-node-name: {"ts":1663845935494.523,"caller":"eviction/eviction_manager.go:254","msg":"Eviction manager: failed to get summary stats","err":"failed to get node info: node \"talos-anonymized-node-name\" not found"}
talos-anonymized-node-name: {"ts":1663845935523.2158,"caller":"kubelet/kubelet.go:2391","msg":"Container runtime network not ready","networkReady":"NetworkReady=false reason:NetworkPluginNotReady message:Network plugin returns error: cni plugin not initialized"}

talos-anonymized-node-name: {"ts":1663845935567.3572,"caller":"kubelet/kubelet.go:2466","msg":"Error getting node","err":"node \"talos-anonymized-node-name\" not found"}
talos-anonymized-node-name: {"ts":1663845935667.9382,"caller":"kubelet/kubelet.go:2466","msg":"Error getting node","err":"node \"talos-anonymized-node-name\" not found"}
talos-anonymized-node-name: {"ts":1663845935768.2312,"caller":"kubelet/kubelet.go:2466","msg":"Error getting node","err":"node \"talos-anonymized-node-name\" not found"}
```
Edit: I missed this other message:
```
talos-anonymized-node-name: {"ts":1663848163959.026,"caller":"certificate/certificate_manager.go:471","msg":"kubernetes.io/kube-apiserver-client-kubelet: Failed while requesting a signed certificate from the control plane: cannot create certificate signing request: Unauthorized\n"}
```

And those three last messages repeated a lot. Does this error mean that `kubelet` doesn't have the proper user certificate or kubeconfig to speak to the current control plane?

```
User "system:anonymous" cannot create resource "nodes" in API group "" at the cluster scope
```

**Reply by smira** [reply]
looks like kubelet got issues wrong certificate? the real error might be somewhere above, but I would at least check that join token matches across Talos config and join tokens in the cluster.

**Reply by ethernet-zero** [reply]
Damn, I think I found it!

My control plane doesn't use encryption for secret data at rest, but the new Talos node does. When generating the configuration files from the secrets, it generated an `aescbcEncryptionSecret`. After copying the encryption config to the control plane nodes and triggering a re-run of their corresponding kube-apiservers, the Talos node finally showed up in the list.

Now on to the next thing: Calico doesn't initialize because the control plane nodes have a `/etc/cni/net.d/calico.conflist.template` file and the Talos node doesn't. I'm now looking into that.

**Reply by ethernet-zero** [reply]
Fixed. I had to add `127.0.0.1` to `cluster.apiServer.certSANs` or else `kube-proxy` would be unable to watch for services:

```
E0923 10:47:52.129067       1 reflector.go:138] k8s.io/client-go/informers/factory.go:134: Failed to watch *v1.Service: failed to list *v1.Service: Get "https://127.0.0.1:6443/api/v1/services?labelSelector=%21service.kubernetes.io%2Fheadless%2C%21service.kubernetes.io%2Fservice-proxy-name&limit=500&resourceVersion=0": x509: certificate is valid for <ENDPOINT_IPS>, not 127.0.0.1
```

Also, I made Calico load that conflist template file from a ConfigMap. I'm not sure if that's a good practice for Calico or not, but it works.

That's it, there's my new and shiny Talos control plane node. Now I can send Kubespray to the deepest reaches of the oblivion where it deserves to be. :laughing:

**Reply by smira** [reply]
awesome! if you have any additions/fixes to the documentation, please create a pull request!

**Reply by ethernet-zero** [reply]
Will do. Thanks again!
